### ISSUE TITLE
Feature Request: Open Folder to the Left/Right

### ISSUE THREAD

**Issue by ThomasSeeker** [ssborbis/ContextSearch-web-ext#569]
Hey there!

This is a feature request.

I use my Menu in View-Mode: Text. 

And if I click on a folder, it opens in a completely new menu-window.

Is there a way to allow the folder-content to open in an additional window left or right to the main menu-window? ... Basically like the native Firefox-Context-Menu does it.

I think, this would allow for a more flowing use of the folders. 

But it is not super important, I must add. ;-)

Cheers and best Greetings,
Thomas

**Comment by 7peanuts** [comment]
Wow, sounds great, currently I am using the mouse to hover to open the folder.
CS Options -> Advanced -> `openFoldersOnHoverTimeout`

**Comment by ThomasSeeker** [comment]
> Options -> Advanced -> `openFoldersOnHoverTimeout`

Yes, I found this option just today! :-) 

And I tried it out. And it is wonderful, ... 

... the only problem is, that if I accidentally rest my mouse on a folder then, it opens and then I sort of get lost in the folder and need to find my way back. ... That gave me the final impulse to write my feature request, because if the folder would open left/right, you wouldn't get lost in it ... you could just move on. ... This would be the best IMO. :-)

The other reason to write this: The deeper down the folders are towards the bottom of the menu, the harder it gets to jump in and out of a folder. And if they only have 2 or 3 entries, you also need to go up and down with the mouse a lot. 

Hey 7peanuts, ... glad, you like it!

**Comment by 7peanuts** [comment]
Yes, you're right, especially if there are subfolders.

**Comment by ssborbis** [comment]
I'm looking into this, but it's actually quite difficult to implement with the current way the quick menu is coded. It could be much simpler, but every menu needs to be inside an iframe to protect users from a snooping website. If I have anything worth testing, I'll post here.

**Comment by ThomasSeeker** [comment]
Wow, thank you!
That would be wonderful!
:-)

**Comment by ssborbis** [comment]
I have a rough draft in the development branch for testing. Enable hidden option `quickMenuUseCascadingFolders`

**Comment by ThomasSeeker** [comment]
Thank you for this, ssborbis!

Unfortunately I am too much of a noob in this area to get the development branch working.

I searched the web and tried these commands:

```
git clone https://github.com/ssborbis/ContextSearch-web-ext.git
cd ContextSearch-web-ext
git branch -a
git checkout development
```

And a new `git branch -a` shows, that it indeed switched to development. ... But when I restart firefox I strongly believe it is still the old version of CS running and the new option is not there. 

What am I missing here?

**Comment by ThomasSeeker** [comment]
Sorry for the fuss!
I found a description in the README-file to "install it as a "temporary addon" from about:debugging -> This Firefox -> Load Temporary addon and browse to src/manifest.json in the unzipped source code"

And I checked the new option! :-)

But now the folders do not open in the QM at all anymore. Neither with 0 (and Click) as the openFoldersOnHoverTimeout or something else.

**Comment by ssborbis** [comment]
> But now the folders do not open in the QM at all anymore. Neither with 0 (and Click) as the openFoldersOnHoverTimeout or something else.

Hmm. odd. I've been testing on the new code all day and haven't had an issue using mouse buttons and open on hover. I might need a copy of your config in case you have some conflicting settings that I do not.

**Comment by ThomasSeeker** [comment]
Wow, that looks good, what you have running there! :-)

Sent the file via email.

**Comment by ssborbis** [comment]
I'll post a beta after I clean a few things up.

**Comment by ssborbis** [comment]
1.46 beta https://github.com/ssborbis/ContextSearch-web-ext/releases/download/v1.46/contextsearch_web_ext-1.46b1.xpi

**Comment by ThomasSeeker** [comment]
For me here the problem persists.

The folders still don't open at all with the new setting "quickMenuUseCascadingFolders": true, .

I does open in one case though: 

First I open the QM with hold--down-right-click.
Second I click on a folder (nothing happens).
Third I make another hold--down-right-click somewhere else on the screen.
...   Now the old QM disappears. The new one (at the new location) appears *AND* the folder content (the wanted cascaded second tier QM-window) appears at the exact location where it should have been with the first QM-opening.

I think this behaviour did already occur with the development version. But I was not sure, because I couldn't reproduce it. (Now (with the above steps) it is reproducible.)

I really hope you find a solution, because it looks damn good on your machine! 

PS: Due to hardware restrictions I still run Firefox 88.0. Could this cause the problem?

**Comment by ssborbis** [comment]
Can you send me your config? I can't reproduce that bug with my settings.

**Comment by ssborbis** [comment]
> PS: Due to hardware restrictions I still run Firefox 88.0

Oof, maybe.  I'll get back to you on that when I have a chance to test.

**Comment by ssborbis** [comment]
@ThomasSeeker 

I tested on Waterfox, and I think I'm seeing a bit of what you are. The folders are actually there, but they have no width or height. I'll get this worked out on Waterfox, and hopefully that will also fix the FF88 issues.

**Comment by ssborbis** [comment]
I'm dropping a new beta Please export your settings before using so you don't have any issues going back to the previous version. You shouldn't need it, but better to be safe.

https://github.com/ssborbis/ContextSearch-web-ext/releases/download/v1.46/contextsearch_web_ext-1.46b2.xpi

**Comment by ThomasSeeker** [comment]
Ah, so beautiful! ... Yippee! It works! 

Thank you so much, ssborbis!

A couple minor things, you probably are already aware of (See in the gif):

1. The folder does not close after one passes it's row. Is this by purpose? It does make it easier to aim, without needing to stay exactly in the row, though.

2. The main window has a little hick-up (and changes it's height) if the folder window is open and one leaves the QM (sometimes only on second try).

3. Only in the rare case, that I open the QM somewhere else again without closing it before that, the folder stays open at the old position. 

4. (maybe not related to this new change): I used to be able to set the width of the QM with `[data-menu="quickmenu"] {width: 230px !important;}`. Now this applies only to the inner rows of the QM, not the whole QM anymore. Is there another CSS-selector for this necessary? I couldn't find it.

I said it before, but I have to say it again: This Addon rocks! :-)

![04](https://user-images.githubusercontent.com/44918701/192761972-6278c64d-5b15-4c3f-a9ba-cf0f19f42c57.gif)

**Comment by ssborbis** [comment]
Some of those issues have already been addressed in the dev branch, and I'll keep an eye out for the others

> (maybe not related to this new change): I used to be able to set the width of the QM with [data-menu="quickmenu"] {width: 230px !important;}. Now this applies only to the inner rows of the QM, not the whole QM anymore. Is there another CSS-selector for this necessary? I couldn't find it.

I made some changes to the styling to fix a few things. The new rule for that particular setting would be:

```css
[data-menu="quickmenu"] {--single-column-width: 230px}
```

**Comment by ssborbis** [comment]
New beta https://github.com/ssborbis/ContextSearch-web-ext/releases/download/v1.46/contextsearch_web_ext-1.46b3.xpi

**Comment by ThomasSeeker** [comment]
Beautiful! ... And the hickup is also gone! :-)

Say, ssborbis, how did you manage to realize this change so fast? ... Because just a couple days ago you said, it would be quite difficult to do.

**Comment by ssborbis** [comment]
> Say, ssborbis, how did you manage to realize this change so fast? ... Because just a couple days ago you said, it would be quite difficult to do.

I'm just that good. Seriously though, there's a ton of integration left to be done. Keep using it and you'll see the bugs come out.

**Comment by ThomasSeeker** [comment]
> I'm just that good.

Yes, obviously! ;-)

> Keep using it and you'll see the bugs come out.

Do you want me to report everything I encounter in this regard, or would this be more annoying than helpful?

Also, if there is another way I can help, please let me know!

**Comment by ssborbis** [comment]
> Do you want me to report everything I encounter in this regard, or would this be more annoying than helpful?

Report them here. I'll take any and all.

**Comment by ThomasSeeker** [comment]
> Report them here. I'll take any and all.

:-D

Ok, a little one would be: If I have folders set to "Cascading" in the settings ... it works fine. ... But if I later open the settings page again, the folder gets automatically set back to "Single". (every time I open the settings page)

**Comment by ssborbis** [comment]
> Ok, a little one would be: If I have folders set to "Cascading" in the settings ... it works fine. ... But if I later open the settings page again, the folder gets automatically set back to "Single". (every time I open the settings page)

That has been fixed. I just need to push the changes. Keep 'em coming.

**Comment by ThomasSeeker** [comment]
Ok, there is not much else to complain about here on my side so far. ;-)

The closing behaviour I in part already wrote about:  The folder-window does not close after I pass the folder in the main-window. 

But on the other hand, IMO it also closes to fast: 1. after I leave the folder-menu to the outside it closes. I'd like to have it stay open in this case. Like the main-window also stays open if I leave it (until click or scroll). 

And also IMO it closes to fast: 2. When I click an entry in the folder-menu, and that leads me to anther page or application, ... when I come back, the folder-menu is gone. Here also, I'd like it to stay open like the main-window.

But I don't know, if the latter ones are just my personal preference or actually make sense for everybody!?

And a last little thing just for the appearance: If I move up and down my menu across several folders the FF-Refresh-Page-Button goes kind of nuts. If there is an easy way to suppress that, I would vote for it! :-)

**Comment by ssborbis** [comment]
> The folder-window does not close after I pass the folder in the main-window.

That is somewhat intentional. Things got weird when using grid mode. I may look into that in the future, but for now, not a huge deal

> But on the other hand, IMO it also closes to fast: 1. after I leave the folder-menu to the outside it closes. I'd like to have it stay open in this case. Like the main-window also stays open if I leave it (until click or scroll).

> When I click an entry in the folder-menu, and that leads me to anther page or application, ... when I come back, the folder-menu is gone. Here also, I'd like it to stay open like the main-window.

These may be the same thing. I'll add an option

> And a last little thing just for the appearance: If I move up and down my menu across several folders the FF-Refresh-Page-Button goes kind of nuts. If there is an easy way to suppress that, I would vote for it! :-)

That may be fixed already, since I'm not seeing it.

**Comment by ThomasSeeker** [comment]
> > The folder-window does not close after I pass the folder in the main-window.
> 
> That is somewhat intentional. Things got weird when using grid mode. I may look into that in the future, but for now, not a huge deal

In my case, about 90% of the time I only use the first tier QM. In the second tier I have only the less used entries. ... But almost every time I open QM and move to the wanted first tier entry, some folder opens and stays open and kind of distracts attention. ... So if it is difficult to change, ... would it be possible, to make the cascading view respect the openFoldersOnHoverTimeout setting of 0 ... to force me to click a folder to open it? ... so they would stay closed, if they are not needed?

> 
> > But on the other hand, IMO it also closes to fast ...
> 
> These may be the same thing. I'll add an option
> 
Ah, that sounds good! ... Looking forward to this! :-)

**Comment by ssborbis** [comment]
> would it be possible, to make the cascading view respect the openFoldersOnHoverTimeout setting of 0 ... to force me to click a folder to open it?

I'll need to add another setting, but easily done.

I'm trying to get a new beta / release going, but I'm waiting on Mozilla to fix a bug preventing me from submitting this addon for validation and signing.

**Comment by ThomasSeeker** [comment]
Ah, wonderful, ssborbis!

Thank you very much!

:-)

**Comment by ssborbis** [comment]
v1.46 is live. Open a new issue for any bugs regarding cascading folders
